state/validation.go

```diff
@@ -42,6 +42,12 @@ func validateBlock(stateDB dbm.DB, state State, block *types.Block) error {
 		return fmt.Errorf("Wrong Block.Header.TotalTxs. Expected %v, got %v", state.LastBlockTotalTx+newTxs, block.TotalTxs)
 	}
 
+	// Don't risk recomputing, since marshalling may be slow.
+	blockSize := block.Size()
+	if blockSize > state.ConsensusParams.BlockSize.MaxBytes {
+		return fmt.Errorf("Block is too large. Got %d, maximum is %d", blockSize, state.ConsensusParams.BlockSize.MaxBytes)
```

Comment:
Is this the correct place to ensure that the received block isn't too large? It would be more ideal to ensure this before decoding the block.

Reply:
The race conditions this causes makes it clear that this is not the correct place to check this.

Reply:
melekes pointed out that this already ensured here: https://github.com/tendermint/tendermint/blob/dde96b75ce3684283f980b968b9edb29ecdff973/consensus/state.go#L1420